### DISCUSSION TITLE
What's going on under the hood?

### DISCUSSION THREAD

**Discussion by rwdr0** [guidance-ai/guidance#870] [Q&A]
I am trying to understand how Guidance is able to _guide_ a language model. Can someone please explain to me the implementation details or point me to relevant resources?

**Comment by Harsha-Nori** [comment]
Happy to expand on more details, but as a first step, you might find this talk I gave recently helpful: https://www.youtube.com/watch?v=qXMNPVVlCMs

**Reply by rwdr0** [reply]
Thank You- It was a great watch! 

It's fascinating how manipulating the actual inference engine is proving to be a viable steering strategy in addition to the long standing traditional approach of fine-tuning models.

And yes on that note- how exactly is Guidance able to hook into a model's inference layer during runtime?

**Reply by Harsha-Nori** [reply]
Glad you enjoyed the talk! Totally agree that steering at inference is super interesting :)! For all local models (e.g. anything hosted on `Transformers` or `LlamaCpp`), we directly modify the inference stack locally. You can see examples of our integrations here: https://github.com/guidance-ai/guidance/blob/e234c565b61ffb90dbbf81cd937a00505ef79649/guidance/models/transformers/_transformers.py#L139

For most remote models (e.g. Anthropic/Gemini/OpenAI Models), we're limited in how much we can manipulate the inference stack by the APIs the model builders expose for us. However -- as mentioned in the video -- we're actively working on bringing guidance support to the cloud :)
